File: MatrixExp.py
import numpy as np
from scipy.linalg import expm #used for compare my results with scipy results
from time import time

#calculate the norm of matrix and its do that :norm of infinite is abs all elements then sum all the rows of matrix and get max sum of row 
def norm_Mat(A):
    return np.max(np.sum(np.abs(A), axis=1)) #using norm of maximum 

#this method calculate the שארית of malorin/lagrang (the n is the num of iteration should tor taylor do)
def calculate_Rn(A,epsilon):
    if np.isscalar(A): #check if number is scalar and not matrix or vector
        return 0
    norm_M = norm_Mat(A)
    e_norm_M = np.exp(norm_M)  
    n=0
    term = e_norm_M * norm_M
    while(term>epsilon):  #while the Rn small than epsilon continue to add to n 
        n=n+1
        term = term*(norm_M*(1/(n+1)))
    return n 

#implement method for calculate multiply between matrixes 
def multiMatrix(A,oldResult):
    result = np.dot(A,oldResult)  #multiply between two matrixes (old Result is A^n-1 * A (instead of calculate every time from start)))
    return result

#return the result multiply of two matrix
def cal_exponent_mat(A,result):
    multiplyMat = multiMatrix(A,result).astype(np.float64)
    return multiplyMat.astype(np.float64)

#sum the tor of taylor
############ the code work for real numbers not for complex. it work for complex just for easily change np.float64 to np.complex128
def sum_tor(A,num_iterations):
    if np.isscalar(A): #i check if the matrix is number then i make a normal exponent
        return np.exp(A)
    dim = np.shape(A)[0] #get dim of matrix 
    if(np.shape(A)[0]!=np.shape(A)[1]): #check if the dim of matrix not equal in rows and cols 
        print("to calculate exponential of matrix you need matrix nXn")
        return
    I=np.eye(dim,dim,dtype=np.float64)  #create identy matrix (מטריצת יחידה)
    resultMat = I
    resultMat+=A #add I + A (numpy know to sum them)
    result = A
    for iter in range(2,num_iterations+1): #loop for sum until n (n I get from calculate_Rn func)
        result = cal_exponent_mat(A,result)
        result = result*(1/iter)
        resultMat+= result
    return resultMat    

#this function read a file txt
def read_Mat_TXT():
    with open('inv_matrix(1000x1000).txt', 'r') as file:
        content = file.read()
    rows = content.split('\n')
    matrix = [list(map(float, row.split(','))) for row in rows if row]
    matrix_np = np.array(matrix)
    return matrix_np.astype(np.float64)

#this function write results to file 
def write_result_to_file(filename, result):
    with open(filename, 'w') as file:
        file.write(str(result))


matrix = read_Mat_TXT()

epsilon = 1.5e-12
t = time()
scipyResult = expm(matrix)
t = time() - t
print("scipyTime",t)
print("---------------------")
print("please wait the calcluate in my computer take 1 min to finish!!!")
t = time()
num_iterations = calculate_Rn(matrix,epsilon)
myWayResult = sum_tor(matrix,num_iterations)
t = time() - t
print("myWay Time",t)

#I save the results for files 
write_result_to_file("ScipyResults.txt",scipyResult)
write_result_to_file("MyWayResults.txt",myWayResult)
print("the print of results to txt files finished Successfully")

File: MatrixExp.hs
import Numeric.LinearAlgebra.HMatrix
import Prelude hiding ((<>))
import Data.Time.Clock


--this method split rows of matrix by ch
splitBy :: Char -> String -> [String]
splitBy delimiter = words . map (\c -> if c == delimiter then ' ' else c)

-- this method read matrix from file
readMatFromFile :: FilePath -> IO (Matrix Double)
readMatFromFile filePath = do
    content <- readFile filePath
    let rows = lines content -- divide the content for rows
        matrixList = [map read (splitBy ',' row) | row <- rows] :: [[Double]] -- convert each row for list
        numRows = length matrixList
        numCols = length (head matrixList)
    return $ (numRows><numCols) (concat matrixList) -- build a matrix from list of lists


--this method calculate the norm of matrix (maximum norm that convert each element to plus instead of minus (abs) then sum each row in matrix then return the maximum row
norm_Mat::Matrix Double -> Double
norm_Mat m = maximum $ map sum $ toLists $ cmap abs m

--this method return the n from calculate the Rn of maclorin/lagrang (שארית מקלורן כדי לדעת כמה איטרציות לבצע בטור טיילור בחישוב האקספוננט)
calculate_Rn::Matrix Double -> Double -> Int --get a Matrix + epsilon and return a n (n is the num of iterations that i will give to sum_tor_taylor)
calculate_Rn m epsilon = 
 let norm_M = norm_Mat m
     e_norm_M = exp norm_M
     term = norm_M*e_norm_M
  in check_greater_than_epsilon 1 term epsilon norm_M

--this method check if term big than epsilon if yes return this n if not i increse n and check again (check again by recursive call to same function(this function instead of while loop
check_greater_than_epsilon::Int -> Double -> Double -> Double -> Int
check_greater_than_epsilon n term epsilon norm_M | term < epsilon = n-1
                                                 | otherwise = let
                                                                calcNewTerm = term * ((norm_M)*1/(fromIntegral (n + 1)))
                                                               in check_greater_than_epsilon (n+1) calcNewTerm epsilon norm_M

--this method calculate the tor of taylor from 0 to num_iterations 
sum_tor_taylor:: Matrix Double -> Int -> Matrix Double
sum_tor_taylor m num_iterations = 
 let dim = size $ m --get demonosial of matrix 
     rowSize = rows $ m
     identMat = ident rowSize --build ident matrix
     resultMat = identMat + m
     result = m
  in sum_matrix_for_tor_taylor m num_iterations 2 result resultMat
 

--this method make the sum of matrixes in recursive way,after it finished return the value to sum_tor_taylor
sum_matrix_for_tor_taylor:: Matrix Double -> Int -> Int -> Matrix Double -> Matrix Double -> Matrix Double
sum_matrix_for_tor_taylor m 0 fracVal result resultMat = resultMat
sum_matrix_for_tor_taylor m num_iterations fracVal result resultMat = 
 let myresult = multiMatrix m result
     mulFracResult = myresult*(1/(fromIntegral(fracVal)))
     newresultMat = resultMat + mulFracResult --like numpy the hmatrix can add two matrixes by + operator
  in sum_matrix_for_tor_taylor m (num_iterations-1) (fracVal+1) mulFracResult newresultMat


--this method multiply two matrix (like numpy.dot in python)
multiMatrix::Matrix Double -> Matrix Double -> Matrix Double
multiMatrix m resultMat = resultMat <> m --this like NUMPY.DOT in python 

--this method write the matrix to a file
writeResultToFileSimple :: FilePath -> Matrix Double -> IO ()
writeResultToFileSimple filename result = writeFile filename (show result)

main :: IO ()
main = do
     let input_file = "inv_matrix(1000x1000).txt"
     m <- readMatFromFile input_file
     let epsilon = 1.5e-12
     print("please wait the calcluate in my computer take 1 min to finish!!!")
     tt_start <- getCurrentTime
     let num_iterations = calculate_Rn m epsilon
     let res = sum_tor_taylor m num_iterations
     let firstElementInMyMat = res ! 0 ! 0 --get first element from matrix because haskell is lazy language and if we dont need to print then it will not do the calculate
     print $ firstElementInMyMat
     tt_end <- getCurrentTime
     print ("myWay Time:")
     print (diffUTCTime tt_end tt_start)
     print("---------------")
     tt_start <- getCurrentTime
     let hmatRes = expm m
     let firstElementInHMATRIX = hmatRes ! 0 ! 0 --get first element from matrix because haskell is lazy language and if we dont need to print then it will not do the calculate
     print $ firstElementInHMATRIX
     tt_end <- getCurrentTime
     print ("HMatrixWay Time:")
     print (diffUTCTime tt_end tt_start)
     print("please wait to print the results to TXT files!!!")
     writeResultToFileSimple "MyWayResults_Haskell.txt" res   --write the results for txt file
     writeResultToFileSimple "HmatrixResults_Haskell.txt" hmatRes
     print("the print finished Successfully")



